torchao/prototype/scaled_grouped_mm/scaled_grouped_mm.py

```diff
@@ -0,0 +1,415 @@
+            A,
+            float8_config.cast_config_input.target_dtype,
+            scaling_granularity=float8_config.cast_config_input.scaling_granularity,
+            axiswise_dim=-1,
+            round_scales_to_power_of_2=float8_config.round_scales_to_power_of_2,
+        )
+        A_scaled = A.to(torch.float32) * A_scales
+        A_fp8_row_major = to_fp8_saturated(
+            A_scaled, float8_config.cast_config_input.target_dtype
+        )
+
+        # Convert B to float8, column-major for right operand of grouped GEMM.
+        # B shape: (B, K, N)
+        # B scales must be computed rowwise keeping the outer/final dim, so:
+        # B_scales shape: (B, 1, N)
+        B_scales = tensor_to_scale(
+            B_col_major,
+            float8_config.cast_config_weight.target_dtype,
+            scaling_granularity=float8_config.cast_config_weight.scaling_granularity,
+            axiswise_dim=-2,
+            round_scales_to_power_of_2=float8_config.round_scales_to_power_of_2,
+        )
+        B_scaled = B.to(torch.float32) * B_scales
+        B_fp8_col_major = to_fp8_saturated(
+            B_scaled, float8_config.cast_config_weight.target_dtype
+        )
+
+        # Perform scaled grouped GEMM and return result.
+        # output shape: scaled grouped mm of (M,K) @ (B,K,N) = (M,N)
+        return torch._scaled_grouped_mm(
+            A_fp8_row_major,
+            B_fp8_col_major,
+            A_scales.squeeze().reciprocal(),
+            B_scales.squeeze().reciprocal(),
+            offs,
+            out_dtype=out_dtype,
+            use_fast_accum=float8_config.gemm_config_output.use_fast_accum,
+        )
+
+    @staticmethod
+    def backward(ctx, grad_output: torch.Tensor):
+        A, B, offs = ctx.saved_tensors
+        float8_config = ctx.float8_config
+        out_dtype = ctx.out_dtype
+
+        # Convert grad_output to float8, row-major for left operand of grouped GEMM
+        # needed for grad_A: grad_output @ B
+        #
+        # grad_output shape: (M, N)
+        # grad_output_scale shape: (M, 1)
+        grad_output_scales = tensor_to_scale(
+            grad_output,
+            float8_config.cast_config_grad_output.target_dtype,
+            scaling_granularity=float8_config.cast_config_grad_output.scaling_granularity,
+            axiswise_dim=-1,
+            round_scales_to_power_of_2=float8_config.round_scales_to_power_of_2,
+        )
+        grad_output_scaled = grad_output.to(torch.float32) * grad_output_scales
+        grad_output_fp8_row_major = to_fp8_saturated(
+            grad_output_scaled, float8_config.cast_config_grad_output.target_dtype
+        )
+
+        # Convert B to non-transposed, float8, column-major for right operand of grouped GEMM
+        # needed for grad_A: grad_output @ B.
+        # Since B was transposed before entry to forward, we need to transpose it back here for this.
+        B_non_transposed_col_major = B.contiguous().transpose(-2, -1)
```

Comment:
It might be better to scale and transpose B in forward, and store only quantized version (to minimize memory)

Reply:
good idea, done. hopefully torch.compile can do some fusion here and read B once and write both outputs simultaneously (float transposed column major, float8 non-transposed column major).